### PR TITLE
gateway controller: consolidate with waypoint controller

### PR DESCRIPTION
**Pull request by howardjohn**
Currently, there is a gateway deployment controller and a waypoint controller. These serve the same purpose: convert a Gateway to a {Deployment,ServiceAccount,etc}, and share most of the same code.

Today, the ingress mode uses a minimal deployment and relies on injection to fill the rest. This was always intended as a short term process until we had a reasonable upgrade mechanism. Since then, Mitch has implemented the new upgrade mechanism in [doc](https://docs.google.com/document/d/1nQuf43sMvPyqNI8MKNTzi13pUn_WDqLULT55z5s2duY/edit#), which also works for gateways.

This PR merges the two code bases, folding the old waypoint controller into deployment controller.

Changes:
* MAJOR: ingress Gateways are now fully rendered Deployments and do not use injection at all. The actual Pod spec created is the same, but the mechanism for upgrades is now different (deployment will be updated when config changes -> standard k8s rolling restart, rather than injection config changes -> user manually rolling restarts).
* Internal: both types now use YAML templates from the standard template configmap, allowing them to be update at runtime as needed.
* Internal: templates are now just arbitrary (ordered) templates of resources to apply; if we wanted we could easily add a new type in there (such as HPA, etc) without code changes.
* Internal: now that injection isn't used, the actual Deployment template needs to contain all the info that was previously injected, so it does...
* Internal: waypoints previously had a very static template; this is now a bit more dynamic to mirror ingress since we have the full values.yaml etc input available. Fixes https://github.com/istio/istio/issues/43258
* Minor: Waypoints now handle gateway class correctly (inherited from the ingress existing code), fixes https://github.com/istio/istio/issues/43264
* Minor: a GatewayClass for waypoint is automatically registered to the cluster

### CODE REVIEW

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 1]
```diff
@@ -187,12 +276,12 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
 	gc, _ := d.gatewayClassLister.Get(string(gw.Spec.GatewayClassName))
 	if gc != nil {
 		// We found the gateway class, but we do not implement it. Skip
-		if gc.Spec.ControllerName != ControllerName {
+		if !knownControllers.Contains(string(gc.Spec.ControllerName)) {
```
Comment (bleggett) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:279]: nit: is this worth a debug log?

**Review by bleggett — APPROVED** [review]

**Comment by linsun** [comment]
LGTM - no major concerns for ambient. Only concern is this could introduce quite some change to our docs on gateway deployment when this is merged upstream, IIUC.

**Comment by linsun** [comment]
Will wait till @therealmitchconnors has a chance to review

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 2]
```diff
@@ -297,20 +369,54 @@ func (d *DeploymentController) RenderServiceAccountApply(input MergedInput) *cor
...
+	ProxyImage  string
+	ProxyConfig *meshapi.ProxyConfig
+	MeshConfig  *meshapi.MeshConfig
+	Values      map[string]any
+}
+
+func (d *DeploymentController) render(templateName string, mi TemplateInput) ([]string, error) {
+	cfg := d.injectConfig()
+
+	// TODO watch for template changes, update the Deployment if it does
```
Comment (therealmitchconnors) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:385]: I believe this should have been resolved as a part of my PR.  Sorry for hanging reference.

File: pilot/pkg/config/kube/gateway/gatewayclass.go [hunk 3]
```diff
@@ -16,16 +16,20 @@ package gateway
...
 	gatewayclient "sigs.k8s.io/gateway-api/pkg/client/clientset/versioned/typed/apis/v1beta1"
 	lister "sigs.k8s.io/gateway-api/pkg/client/listers/apis/v1beta1"
 
+	"istio.io/istio/pilot/pkg/model/kstatus"
 	"istio.io/istio/pkg/kube"
 	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/util/istiomultierror"
 )
 
 // ClassController is a controller that creates the default Istio GatewayClass. This will not
```
Comment (therealmitchconnors) [inline pilot/pkg/config/kube/gateway/gatewayclass.go:35]: This comment needs updating to reflect that we are now syncing multiple classes.

**Review by therealmitchconnors — COMMENTED** [review]
Minor comments here and there, but overall LGTM

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 4]
```diff
@@ -91,19 +88,101 @@ type DeploymentController struct {
...
+					Message: "Deployed gateway to the cluster",
+				},
+			}
+		},
+	},
+	constants.WaypointGatewayClassName: {
+		controller: constants.ManagedGatewayMeshController,
+		templates:  "waypoint",
+		enabled: func(gw *gateway.Gateway) bool {
+			// we manage all "mesh" gateways
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:138]: Yes lets do that in another followup

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 5]
```diff
@@ -91,19 +88,101 @@ type DeploymentController struct {
...
+	for _, v := range classInfos {
+		res.Insert(v.controller)
+	}
+	return res
+}()
+
 // NewDeploymentController constructs a DeploymentController and registers required informers.
 // The controller will not start until Run() is called.
-func NewDeploymentController(client kube.Client) *DeploymentController {
+func NewDeploymentController(client kube.Client, webhookConfig func() inject.WebhookConfig, injectionHandler func(fn func())) *DeploymentController {
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:181]: Will look into it later maybe?

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 6]
```diff
@@ -204,53 +293,46 @@ func (d *DeploymentController) Reconcile(req types.NamespacedName) error {
 func (d *DeploymentController) configureIstioGateway(log *istiolog.Scope, gw gateway.Gateway) error {
 	// If user explicitly sets addresses, we are assuming they are pointing to an existing deployment.
 	// We will not manage it in this case
-	if !IsManaged(&gw.Spec) {
+	gi, f := classInfos[string(gw.Spec.GatewayClassName)]
+	if !f {
+		return nil
+	}
+	if !gi.enabled(&gw) {
 		log.Debug("skip unmanaged gateway")
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/deploymentcontroller.go:301]: good call, just disabled

**Review by linsun — APPROVED** [review]
Thanks Mitch! I don't see any major concern from you and John pinged me that he will address your comments in a follow up PR.

**Pull request merged**